### ISSUE TITLE
Native BAM parsing [feature]

### ISSUE THREAD

**Issue by flaar94** [biopython/biopython#4868]
I was wondering if you think it would be worth adding support for native parsing of BAM files? 

From what I can find, the support for BAM files primarily comes from a BGZF reader together with some CLI wrappers from SAM tools.  I think it might be worth it since it would be nice to have a direct way of translating data into Python objects (plus I'm biased because SAM tools and Windows don't play nicely).

Below is an example prototype of a fast-and-dirty approach which more-or-less uses the bgzf reader, translates BAM alignments into SAM lines, then passes it through the SAM reader, though there is a bit a friction coming from the abstract class trying to enforce specific stream types and assuming it has a file object. It also ignores any aux data because I don't see it on the SAM alignment spec, but maybe it's supposed to go in the header

```
from Bio import bgzf
from io import StringIO, BytesIO
from typing import NamedTuple
import struct
import numpy as np
from Bio.Align import sam

bam_path = r"D:\genomics_data\HG00138\sex_chrom\HG00138.bam"

def to_uint(x):
    return int.from_bytes(x, byteorder="little", signed=False)


class ReferenceMetadata(NamedTuple):
    name: str
    length: int


class AlignmentInfo(NamedTuple):
    ref_id: int
    pos: int
    l_read_name: int
    mapq: int
    bin: int
    n_cigar_op: int
    flag: int
    l_seq: int
    next_ref_id: int
    next_pos: int
    tlen: int


class BAMToSAMReader(sam.AlignmentIterator):
    seq_encoding = bytes.maketrans(bytes(range(16)), b"=ACMGRSVTWYHKDBN")
    cigar_encoding = bytes.maketrans(b"012345678", b"MIDNSHP=X")

    qual_mapping = bytes(
        (
            letter + 33
            if 0 <= letter < 94
            else 255
        )
        for letter in range(256)
    )

    def __init__(self, bam_path):
        handle = bgzf.BgzfReader(bam_path, mode="rb")
        self.mode = "b"
        super().__init__(handle)
        # This is set after super().__init__() to avoid issues with parent classes assuming stream is a file object
        self._base_handle = handle
        self._stream = self._sam_line_iter()

    def __iter__(self):
        """Overwrite this to stop it from trying to rewind"""
        return self

    def _read_header(self, stream):
        magic_bam = stream.read(4)
        assert magic_bam == b'BAM\x01', f"{magic_bam} != BAM\x01"
        header_length = to_uint(stream.read(4))
        # Most header lines must be ascii, but the following can be UTF-8:
        # DS for @SQ, @RG, and @PG
        # CL for @PG
        # @CO generally
        header = StringIO(stream.read(header_length).decode("UTF-8"))
        super()._read_header(header)  # This will set self._target_indices

        num_refs = to_uint(stream.read(4))
        self._references = []
        for i in range(num_refs):
            ref_name_length = to_uint(stream.read(4))

            ref_name = stream.read(ref_name_length).rstrip(b"\00").decode("ASCII")  # This should be ascii?
            ref_length = to_uint(stream.read(4))
            self._references.append(ReferenceMetadata(ref_name, ref_length))

    @classmethod
    def seq_nibble_decode(cls, seq: bytes) -> bytes:
        seq_array = np.frombuffer(seq, dtype="<B")

        # split into high and low nibbles (this could probably be done faster at a low level if necessary)
        low, high = seq_array & 0x0F, seq_array >> 4

        # Interweave the high and low nibbles
        interweaved = np.ravel(np.column_stack((low, high)))
        return interweaved.tobytes().translate(cls.seq_encoding)

    @classmethod
    def cigar_nibble_decode(cls, cigar_bytes):
        cigar_arr = np.frombuffer(cigar_bytes, dtype="<I")

        # Split into high and low nibbles
        op, op_len = cigar_arr & 0x0F, cigar_arr >> 4

        # Convert to SAM operation characters
        translated_op = np.strings.translate(op.astype(np.bytes_), cls.cigar_encoding)
        # interweave op_len and op
        cigar = b"".join(np.ravel(np.column_stack((op_len, translated_op))))
        return cigar

    def _sam_line_iter(self):
        while True:
            # Total size of alignment block
            block_size_bytes = self._base_handle.read(4)
            if not block_size_bytes:
                raise StopIteration
            block_size = to_uint(block_size_bytes)

            block = BytesIO(self._base_handle.read(block_size))
            alignment_info = AlignmentInfo._make(struct.unpack("<iiBBHHHIiii", block.read(32)))
            read_name = block.read(alignment_info.l_read_name).rstrip(b"\00")
            # TODO: handle case when there are too many operations
            cigar = self.cigar_nibble_decode(block.read(alignment_info.n_cigar_op * 4))

            seq = self.seq_nibble_decode(
                block.read((alignment_info.l_seq + 1) // 2)) if alignment_info.l_seq > 0 else b"*"
            qual = block.read(alignment_info.l_seq) if alignment_info.l_seq > 0 else b"*"

            alignment_ref_name = self._references[alignment_info.ref_id].name if alignment_info.ref_id != -1 else "*"
            next_ref_name = self._references[
                alignment_info.next_ref_id].name if alignment_info.next_ref_id != -1 else "*"
            # aux = block.read()

            sam_line = "\t".join([
                read_name.decode(),
                str(alignment_info.flag),
                str(alignment_ref_name),
                str(alignment_info.pos + 1),
                str(alignment_info.mapq),
                cigar.decode(),
                next_ref_name,
                str(alignment_info.next_pos + 1),
                str(alignment_info.tlen),
                seq.translate(self.seq_encoding).decode(),
                qual.translate(self.qual_mapping).decode()
            ])
            yield sam_line


for i, x in enumerate(BAMToSAMReader(bam_path)):
    pass
```
and here's the correspondence between alignment data as far as I can tell:
![image](https://github.com/user-attachments/assets/4643435f-601e-4d9e-bccd-706713472719)

**Comment by mdehoon** [comment]
I think that would be useful. Ideally, we would want a `bam` module in `Bio.Align` that implements a `AlignmentIterator` class converting BAM data to `Alignment` objects.

**Comment by peterjc** [comment]
This has been on my mental todo list for a while. I had pure Python BAM reading code on this branch, but that will need reworking for the new ``Alignment`` objects:

https://github.com/peterjc/biopython/tree/seqio-sam-bam

https://github.com/biopython/biopython/compare/master...peterjc:biopython:seqio-sam-bam#diff-07c082d7557600c02ae2ebc339bb136a7ca0817938730c212a54d13a63751f6f

This builds on the BGZF support already in Biopython, much like your outline does.

Note my code hadn't been updated to handle extra long CIGAR strings (which use a bit of the FLAG field as I recall).

**Comment by flaar94** [comment]
> This has been on my mental todo list for a while. I had pure Python BAM reading code on this branch, but that will need reworking for the new `Alignment` objects:
> 
> https://github.com/peterjc/biopython/tree/seqio-sam-bam
> 
> [master...peterjc:biopython:seqio-sam-bam#diff-07c082d7557600c02ae2ebc339bb136a7ca0817938730c212a54d13a63751f6f](https://github.com/biopython/biopython/compare/master...peterjc:biopython:seqio-sam-bam#diff-07c082d7557600c02ae2ebc339bb136a7ca0817938730c212a54d13a63751f6f)
> 
> This builds on the BGZF support already in Biopython, much like your outline does.
> 
> Note my code hadn't been updated to handle extra long CIGAR strings (which use a bit of the FLAG field as I recall).

Oh yeah, that looks quite similar in a lot of ways, but with more checks and seqIO directed instead of alignIO. I think I can probably incorporate those checks without too much trouble. Ideally I could avoid needing the SAM parser and the translation layer (because that will slow things down), but honestly the structure of the SAM parser code kind of befuddles me, so it would be difficult for me to figure out if I'm doing all the same quality control. Also the header appears to be identical between the two file types as well, so that code at least should be shared

I also deferred handling the super large CIGAR strings, I guess they stick it in a tag. Here's what the spec says:

> 4.2.2 N CIGAR OP field
> With 16 bits, n cigar op can keep at most 65535 CIGAR operations in BAM files. For an alignment with
> more CIGAR operations, BAM stores the real CIGAR, encoded the same way as the cigar field in BAM, in
> the CG optional tag of type ‘B,I’, and sets CIGAR to ‘kSmN’ as a placeholder, where ‘k’ equals l seq, ‘m’ is the
> reference sequence length in the alignment, and ‘S’ and ‘N’ are the soft-clipping and reference-clip CIGAR
> operators, respectively

How common is it to have that many operations? The example BAM files I have usually just have a handful of operations. I guess if it's just the same format but in a different place it wouldn't be too hard to parse, though it could slow things down having to loop through the variable length tags

**Comment by peterjc** [comment]
I don't think SAM/BAM really fits under ``Bio.SeqIO``, under ``Bio.Align`` is a better match. So don't read too much into my old code on that branch.

The long CIGAR feature is needed for long read data, Oxford Nanopore or PacBio. It wouldn't make sense for Illumina or even Roche 454 reads as they are far shorter than the 16 bit limit, even with a very bad forced alignment.
